File: thread_pool.cc
#include <chrono>
#include <condition_variable>
#include <cstdlib>
#include <functional>
#include <future>
#include <iostream>
#include <mutex>
#include <queue>
#include <thread>
#include <vector>

/**
 * 线程池：
 *   1. 固定 worker
 *   2. 提交函数、参数
 *   3. 依此返回运行结果
 *
 */

std::mutex cout_mtx;

class ThreadPool {
   public:
    ThreadPool(size_t size) {
        terminated = false;
        pool_size = std::min(static_cast<size_t>(std::thread::hardware_concurrency()), size);
        threads.reserve(pool_size);
        for (size_t i = 0; i < pool_size; i++) {
            threads.push_back(std::thread([i, this]() {
                while (true) {
                    std::function<void()> task;
                    {
                        std::unique_lock<std::mutex> lock(mtx);
                        while (tasks.size() == 0) {
                            if (terminated) {
                                return;
                            }
                            cv.wait(lock);
                        }
                        task = tasks.front();
                        tasks.pop();
                    }
                    task();
                }
            }));
        }
    }

    template <class F, class... Args>
    auto submit(F&& f, Args&&... args) -> std::future<typename std::result_of<F(Args...)>::type> {
        using return_type = typename std::result_of<F(Args...)>::type;
        auto task = std::make_shared<std::packaged_task<return_type()>>(
            std::bind(std::forward<F>(f), std::forward<Args>(args)...));
        auto fut = task->get_future();

        {
            std::unique_lock<std::mutex> lock(mtx);
            if (terminated) {
                throw std::runtime_error("Thread pool has been terminated!");
            }
            tasks.push([task]() -> void { (*task)(); });
        }
        cv.notify_one();
        return fut;
    }

    void close() {
        {
            std::unique_lock<std::mutex> lock(mtx);
            terminated = true;
        }
        cv.notify_all();
        for (auto& t : threads) {
            t.join();
        }
    }

   private:
    size_t pool_size;
    std::vector<std::thread> threads;
    std::queue<std::function<void()>> tasks;
    std::mutex mtx;
    std::condition_variable cv;
    bool terminated;
};

int foo(int arg) {
    std::this_thread::sleep_for(std::chrono::milliseconds(500 + std::rand() % 1000));
    return arg * arg;
}

int g(int a, int b) {
    std::this_thread::sleep_for(std::chrono::milliseconds(500 + std::rand() % 1000));
    return a * b;
}

int main() {
    ThreadPool tp(32);
    std::vector<std::future<int>> ans;
    for (int i = 0; i < 128; i++) {
        ans.push_back(tp.submit(foo, i));
    }
    for (int i = 0; i < 128; i++) {
        ans.push_back(tp.submit(g, i, 128 - i));
    }

    tp.close();
    for (auto& i : ans) {
        std::cout << i.get() << "\n";
    }
}

File: thread_pool.py
from threading import Thread, Lock, Condition
import time


class Result:
    def __init__(self) -> None:
        self.lock = Lock()
        self.value = None
        self.ready = False

    def get(self):
        while True:
            with self.lock:
                if self.ready:
                    return self.value
            time.sleep(0.5)

class ThreadPool:
    def __init__(self, size: int) -> None:
        self.size = size
        self._tasks = []
        self._threads = [Thread(target=self._worker) for _ in range(size)]
        self._cond = Condition(Lock())
        self._running = True
        
        for i in self._threads:
            i.start()

    def _worker(self):
        while True:
            with self._cond:
                while len(self._tasks) == 0:
                    if not self._running:
                        return
                    self._cond.wait()
                func, args, kwargs, result = self._tasks.pop(0)
            val = func(*args, **kwargs)
            with result.lock:
                result.value = val
                result.ready = True

    def submit(self, func, *args, **kwargs) -> Result:
        with self._cond:
            result = Result()
            self._tasks.append((func, args, kwargs, result))
            self._cond.notify()
            return result

    def exit(self):
        with self._cond:
            self._running = False
            self._cond.notify_all()
        for i in self._threads:
            i.join()


if __name__ == "__main__":
    tp = ThreadPool(2)

    def foo(a=None):
        print(a, f' {time.ctime()}')
        time.sleep(1)
        return a
    ans = []
    for _ in range(5):
        ans.append(tp.submit(foo))
    for i in range(5):
        ans.append(tp.submit(foo, i))
    tp.exit()
    for i in ans:
        print(i.get())
